### PR TITLE
Feature #2772 python embedding json

### LINKED ISSUE #2772: Switch from writing temporary Python files in NetCDF to JSON and NumPy serialization
## Describe the Enhancement ##
This idea was discussed during the METplus Engineering meeting on 12/19/23. @jprestop has been having challenges installing MET-12.0.0-beta2 on HPC machines. In particular, there is a runtime linker issue with NetCDF/HDF5 libraries. The MET C++ code requires and links to NetCDF. In addition, the version of Python with which MET is compiled is required to support the NetCDF Python package since that's what is used for writing temporary NetCDF files when the $MET_PYTHON_EXE environment variable is set.

These two NetCDF shared object library files can come into conflict. And requiring them to be identical, i.e. compiled by the same compiler family/version is unrealistic. It may be useful to REMOVE that Python NetCDF requirement. We could try doing so by switching to writing temporary Python files in JSON instead. Note that this is actually how its already being handled in the Python embedding used for TC-Diag.

This is an idea that requires more research to determine whether or not its feasible. Recommend testing by modifying the `write_tmp_dataplane.py` and `read_tmp_dataplane.py` scripts to use JSON instead of NetCDF. The real test is how performant JSON will be compared to NetCDF.

It is not clear whether or not this will actually improve the linker issues that @jprestop has encountered.

### Time Estimate ###
*Estimate the amount of work required here.*
*Issues should represent approximately 1 to 3 days of work.*

### Sub-Issues ###
Consider breaking the enhancement down into sub-issues.
- [ ] *Add a checkbox for each sub-issue here.*

### Relevant Deadlines ###
*List relevant project deadlines here or state NONE.*

### Funding Source ###
*Define the source of funding and account keys here or state NONE.*
2702691

### Assignee ###
- [ ] Select **engineer(s)** or **no engineer** required
- [ ] Select **scientist(s)** or **no scientist** required

### Labels ###
- [x] Review default **alert** labels
- [x] Select **component(s)**
- [x] Select **priority**
- [x] Select **requestor(s)**

### Milestone and Projects ###
- [x] Select **Milestone** as the next official version or **Backlog of Development Ideas**
- [x] For the next official version, select the **MET-X.Y.Z Development** project

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [x] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdataio](https://github.com/dtcenter/METdataio/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)
No impacts.

## Enhancement Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the issue definition above, including the **Time Estimate** and **Funding Source**.
- [ ] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [ ] Complete the development and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [ ] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)** and **Development** issue
Select: **Milestone** as the next official version
Select: **MET-X.Y.Z Development** project for development toward the next official release
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

### PR DESCRIPTION
**Pull request by hsoh-u**
## Expected Differences ##

The default file format of the temporary files is changed to JSON  (for attributes) and numpy serialization for 2D data from NetCDF. The string type arrays for the point data are saved as JSON, too.

- The netCDF package is not loaded automatically at dataplane.py and point.py
- point_nc.py is separated from point.py 
- Refactoring the MET_BASE/scripts/python/met package
- New environment variables were introduced
  - MET_KEEP_TEMP_FILE: "true" or "yes" to keep the temporary files (both MET and Python embedding)
  - Python embedding only
    - MET_PYTHON_DEBUG: "all" or component name ("dataplane" or "point") to enable log messages from the python embedding, default is disabled
    - MET_PYTHON_TMP_FORMAT: "netcdf" to save the NetCDF format for the temporary file, default is JSON (for attributes) mand numpy serialization to avoid NetCDF library conflict between MET and python3.

- [x] Do these changes introduce new tools, command line arguments, or configuration file options? **[No]**</br>
If **yes**, please describe:</br>

- [x] Do these changes modify the structure of existing or add new output data types (e.g. statistic line types or NetCDF variables)? **[No]**</br>
If **yes**, please describe:</br>

## Pull Request Testing ##

- [x] Describe testing already performed for these changes:</br>

Unit test and the same result

How to check if JSON format or netCDF format is used: by setting MET_PYTHON_TMP_FORMAT and MET_PYTHON_DEBUG:
```
export MET_PYTHON_DEBUG="all"
unset MET_PYTHON_TMP_FORMAT
export PYTHON_GRID='G212'
export MET_PYTHON_EXE="/nrit/ral/met-python3/bin/python3.10"
/d1/personal/hsoh/git/pull_request/MET_feature_2772_python_embedding_json/bin/plot_data_plane \
      PYTHON_NUMPY \
      letter_numpy_grid_name_user_python.ps \
      'name = "/d1/personal/hsoh/git/pull_request/MET_feature_2772_python_embedding_json/share/met/python/examples/read_ascii_numpy_grid.py /d1/projects/MET/MET_test_data/unit_test/python/letter.txt LETTER";' \
      -plot_range 0.0 255.0 \
      -title "Grid Name: 'G212'" \
      -v 1

 PYTHON: Save to a temporary JSON file and a temporary numpy output (dataplane)
 PYTHON: Module:        'met.dataplane'
 PYTHON: User python command:   '/d1/personal/hsoh/git/pull_request/MET_feature_2772_python_embedding_json/share/met/python/pyembed/read_tmp_dataplane.py /tmp/tmp_met_data_1485475_0'
 PYTHON: Read from a temporary JSON file and a temporary numpy output (dataplane)
 


export MET_PYTHON_TMP_FORMAT="netcdf"
export PYTHON_GRID='G212'
export MET_PYTHON_EXE="/nrit/ral/met-python3/bin/python3.10"
/d1/personal/hsoh/git/pull_request/MET_feature_2772_python_embedding_json/bin/plot_data_plane \
      PYTHON_NUMPY \
      letter_numpy_grid_name_user_python.ps \
      'name = "/d1/personal/hsoh/git/pull_request/MET_feature_2772_python_embedding_json/share/met/python/examples/read_ascii_numpy_grid.py /d1/projects/MET/MET_test_data/unit_test/python/letter.txt LETTER";' \
      -plot_range 0.0 255.0 \
      -title "Grid Name: 'G212'" \
      -v 1

 PYTHON: Save to a temporary NetCDF file (dataplane)
 PYTHON: Module:        'met.dataplane'
 PYTHON: User python command:   '/d1/personal/hsoh/git/pull_request/MET_feature_2772_python_embedding_json/share/met/python/pyembed/read_tmp_dataplane.py /tmp/tmp_met_data_1486057_0'
 PYTHON: Read from a temporary NetCDF file (dataplane)


export MET_KEEP_TEMP_FILE="true"
/d1/personal/hsoh/git/pull_request/MET_feature_2772_python_embedding_json/bin/plot_data_plane \
      PYTHON_NUMPY \
      letter_numpy_grid_name_user_python.ps \
      'name = "/d1/personal/hsoh/git/pull_request/MET_feature_2772_python_embedding_json/share/met/python/examples/read_ascii_numpy_grid.py /d1/projects/MET/MET_test_data/unit_test/python/letter.txt LETTER";' \
      -plot_range 0.0 255.0 \
      -title "Grid Name: 'G212'" \
      -v 1

check /tmp directory for temporary files

unset MET_KEEP_TEMP_FILE
/d1/personal/hsoh/git/pull_request/MET_feature_2772_python_embedding_json/bin/plot_data_plane \
      PYTHON_NUMPY \
      letter_numpy_grid_name_user_python.ps \
      'name = "/d1/personal/hsoh/git/pull_request/MET_feature_2772_python_embedding_json/share/met/python/examples/read_ascii_numpy_grid.py /d1/projects/MET/MET_test_data/unit_test/python/letter.txt LETTER";' \
      -plot_range 0.0 255.0 \
      -title "Grid Name: 'G212'" \
      -v 1

check /tmp directory for temporary files

```

- [x] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>

- [x] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[Yes]**

John HG is working on this.

- [x] Do these changes include sufficient testing updates? **[No]**

- [x] Will this PR result in changes to the test suite? **[No]**</br>
If **yes**, describe the new output and/or changes to the existing output:</br>

- [ ] Please complete this pull request review by **[Fill in date]**.</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [x] Review the source issue metadata (required labels, projects, and milestone).
- [x] Complete the PR definition above.
- [x] Ensure the PR title matches the feature or bugfix branch name.
- [x] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)** and **Development** issue
Select: **Milestone** as the version that will include these changes
Select: **Coordinated METplus-X.Y Support** project for bugfix releases or **MET-X.Y.Z Development** project for official releases
- [x] After submitting the PR, select the :gear: icon in the **Development** section of the right hand sidebar. Search for the issue that this PR will close and select it, if it is not already selected.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue and delete your feature or bugfix branch from GitHub.

### CODE REVIEW

**Review by jprestop — COMMENTED** [review]
Thanks for all of your work on this @hsoh-u.  When I was reviewing the source issue metadata, I noticed that there was still an alert for "NEEDS MORE DEFINITION" which I removed, but also for "NEEDS ACCOUNT KEY".  I'm hoping you already know what to charge for your work.  Could you please update the issue with the account key and remove that tag?

I had to modify the xml/unit_python.xml file, adding some `<env>` tags with `<pair><name>MET_PYTHON_EXE</name> <value>&MET_PYTHON_EXE;</value></pair>` in order to get the unit tests to pass on Jet, but finally they all did pass:
```-bash-4.2$ perl perl/unit.pl xml/unit_python.xml |tee ../../output/test_output.txt
TEST: python_numpy_grid_name                            - pass -   2.334 sec
TEST: python_numpy_grid_string                          - pass -   2.359 sec
TEST: python_numpy_grid_data_file                       - pass -   2.320 sec
TEST: python_numpy_plot_data_plane                      - pass -   2.390 sec
TEST: python_xarray_plot_data_plane                     - pass -   2.409 sec
TEST: python_numpy_plot_data_plane_file_type            - pass -   2.390 sec
TEST: python_numpy_mode_mixed                           - pass -   3.697 sec
TEST: python_numpy_mode                                 - pass -   4.838 sec
TEST: python_numpy_grid_stat_mixed                      - pass -   4.176 sec
TEST: python_numpy_grid_stat                            - pass -  10.149 sec
TEST: python_numpy_point_stat                           - pass -   9.487 sec
TEST: python_numpy_wavelet_stat                         - pass -   3.231 sec
TEST: python_numpy_wavelet_stat_mixed                   - pass -   3.698 sec
TEST: python_numpy_regrid_data_plane                    - pass -   1.542 sec
TEST: python_numpy_shift_data_plane                     - pass -   1.395 sec
TEST: python_numpy_shift_data_plane_input_arg           - pass -   1.401 sec
TEST: python_numpy_series_analysis                      - pass -   3.418 sec
TEST: python_numpy_mtd                                  - pass -   1.437 sec
TEST: python_numpy_ensemble_stat                        - pass -   1.383 sec
TEST: python_ascii2nc                                   - pass -   1.171 sec
TEST: python_ascii2nc_varname                           - pass -   1.184 sec
TEST: python_stat_analysis                              - pass -   5.065 sec
TEST: python_numpy_grid_name_user_python                - pass -   2.344 sec
TEST: python_ascii2nc_user_python                       - pass -   1.554 sec
TEST: python_numpy_plot_data_plane_user_python          - pass -   2.399 sec
TEST: python_stat_analysis_user_python                  - pass -   5.858 sec
TEST: python_point2grid_pb2nc_TMP_user_python           - pass -   4.718 sec
TEST: python_plot_point_obs_with_point_data             - pass -   1.605 sec
TEST: python_plot_point_obs_CONFIG                      - pass -  11.492 sec
TEST: python_ensemble_stat_FILE_LIST                    - pass -   5.733 sec
TEST: python_point_stat_GRIB1_NAM_GDAS_WINDS            - pass -  15.278 sec
TEST: python_plot_data_plane_SEMILATLON_ZONAL_MEAN      - pass -   2.469 sec
TEST: python_pcp_combine_SEMILATLON_MERIDIONAL_MEAN     - pass -   2.385 sec
TEST: python_plot_data_plane_SEMILATLON_MERIDIONAL_MEAN - pass -   0.568 sec
```
@JohnHalleyGotway is going to update the documentation in this branch to keep everything together.  Once that is ready for review, I can review that and then approve.  

I greatly appreciate everyone's help to get this working on the various machines where it wasn't.  Thank you so much!

**Comment by JohnHalleyGotway** [comment]
@jprestop and @hsoh-u, I added some documentation about these env vars here:
[MET_KEEP_TEMP_FILE](https://met--2803.org.readthedocs.build/en/2803/Users_Guide/config_options.html#met-keep-temp-file), [MET_PYTHON_DEBUG](https://met--2803.org.readthedocs.build/en/2803/Users_Guide/config_options.html#met-python-debug),  and [MET_PYTHON_TMP_FORMAT](https://met--2803.org.readthedocs.build/en/2803/Users_Guide/config_options.html#met-python-tmp-format). Please feel free to tweak the wording as you see fit.

File: docs/Users_Guide/config_options.rst [hunk 1]
```diff
@@ -494,6 +494,49 @@ Where code is running in a production context, it is worth being familiar with
...
+the default shared :code:`/tmp` directory.
+
+.. _met_python_debug:
+
+MET_PYTHON_DEBUG
+----------------
+
+The MET_PYTHON_DEBUG environment variable can be set to enable debugging log messages
+related to Python embedding. These log messages are disabled by default. The environment
+variable cab be set to a value of :code:`all` for all log messages, :code:`dataplane`
```
Comment (jprestop) [inline docs/Users_Guide/config_options.rst]:
```suggestion
variable can be set to a value of :code:`all` for all log messages, :code:`dataplane`
```

**Review by jprestop — APPROVED** [review]
Thanks @JohnHalleyGotway for writing up the documentation.  It sounded great.  There was one typo ("cab" should have been "can"), which I fixed and committed.  Prior to that change, I verified that all tests passed, except there was one failure in the Check for Differences:
```
ERROR: folder /data/output/met_test_output missing 1 files
    python/pb2nc_TMP.nc 
```
This failure is expected because I removed this unit test.  Once adding MET_PYTHON_EXE, it became a duplicate of another unit test.  

I approve this pull request.

**Pull request merged**